<a href="https://colab.research.google.com/github/leorezendde/PI04_Grupo04/blob/main/Analise_taxa_de_abandono_escolar.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("joaoassaoka/taxa-de-abandono-escolar-por-renda-mdia-brasil")

print("Path to dataset files:", path)

100%|██████████| 563k/563k [00:00<00:00, 43.2MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/joaoassaoka/taxa-de-abandono-escolar-por-renda-mdia-brasil/versions/1


In [3]:
import pandas as pd
import os

# Define the path to the dataset directory based on the previous cell's output
dataset_path = "/root/.cache/kagglehub/datasets/joaoassaoka/taxa-de-abandono-escolar-por-renda-mdia-brasil/versions/1"
csv_file_name = "AbandonoEscolar_RendaMedia_2013_2023.csv"
full_csv_path = os.path.join(dataset_path, csv_file_name)


# Carregar o dataset
df = pd.read_csv(full_csv_path)

# Mostrar as primeiras linhas do dataset e os tipos de dados
display(df.head())
display(df.info())

Ano Unidade_Geografica  Regiao Localizacao Dependencia_Administrativa  \
0  2013              Norte  Brasil       Total                      Total   
1  2013              Norte  Brasil       Total                    Pública   
2  2013              Norte  Brasil       Total                    Privada   
3  2013              Norte  Brasil       Total                  Municipal   
4  2013              Norte  Brasil       Total                    Federal   

    Grupo_de_Abandono  Taxa_Abandono  Renda_Media  
0  Ensino Fundamental            4.0         2252  
1  Ensino Fundamental            4.3         2252  
2  Ensino Fundamental            0.3         2252  
3  Ensino Fundamental            4.3         2252  
4  Ensino Fundamental            0.0         2252

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90160 entries, 0 to 90159
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Ano                         90160 non-null  int64  
 1   Unidade_Geografica          90160 non-null  object 
 2   Regiao                      90160 non-null  object 
 3   Localizacao                 90160 non-null  object 
 4   Dependencia_Administrativa  90160 non-null  object 
 5   Grupo_de_Abandono           90160 non-null  object 
 6   Taxa_Abandono               82112 non-null  float64
 7   Renda_Media                 90160 non-null  int64  
dtypes: float64(1), int64(2), object(5)
memory usage: 5.5+ MB


None

# Objetivo
Analisar o conjunto de dados "/kaggle/input/taxa-de-abandono-escolar-por-renda-mdia-brasil/AbandonoEscolar_RendaMedia_2013_2023.csv" para identificar insights e conhecimento sobre as taxas de abandono escolar no Brasil, focando na relação com a renda média, unidades geográficas, dependência administrativa, localização e diferentes níveis educacionais.

In [ ]:
# Limpeza dos dados
# Tratar os valores ausentes na coluna Taxa_Abandono

missing_values_count = df['Taxa_Abandono'].isnull().sum()
total_rows = len(df)
percentage_missing = (missing_values_count / total_rows) * 100

print(f"Número de valores ausentes em Taxa_Abandono: {missing_values_count}")
print(f"Porcentagem de valores ausentes em Taxa_Abandono: {percentage_missing:.2f}%")

if percentage_missing < 5:
    # Imputar com a mediana
    median_taxa_abandono = df['Taxa_Abandono'].median()
    df['Taxa_Abandono'].fillna(median_taxa_abandono, inplace=True)
    print(f"Valores ausentes imputados com a mediana: {median_taxa_abandono:.2f}")
else:
    # Remover linhas com valores ausentes
    df.dropna(subset=['Taxa_Abandono'], inplace=True)
    print("Linhas com valores ausentes removidas.")

missing_values_after = df['Taxa_Abandono'].isnull().sum()
print(f"Número de valores ausentes em Taxa_Abandono após tratamento: {missing_values_after}")

In [ ]:
# Análise exploratória inicial
# Visualizar a distribuição da taxa de abandono escolar e da renda média, e suas tendências ao longo do tempo para o Brasil.

import matplotlib.pyplot as plt

# Filtrar para 'Brasil'
df_brasil = df[df['Unidade_Geografica'] == 'Brasil'].copy()

# Agrupar por ano e calcular a média para plotar as tendências
df_brasil_yearly = df_brasil.groupby('Ano')[['Taxa_Abandono', 'Renda_Media']].mean().reset_index()

# Criar gráfico de linha para a tendência da Taxa_Abandono
plt.figure(figsize=(10, 6))
plt.plot(df_brasil_yearly['Ano'], df_brasil_yearly['Taxa_Abandono'])
plt.xlabel('Ano')
plt.ylabel('Taxa de Abandono (%)')
plt.title('Tendência da Taxa de Abandono Escolar no Brasil (2013-2023)')
plt.grid(True)
plt.show()

# Criar gráfico de linha para a tendência da Renda_Media
plt.figure(figsize=(10, 6))
plt.plot(df_brasil_yearly['Ano'], df_brasil_yearly['Renda_Media'], color='orange')
plt.xlabel('Ano')
plt.ylabel('Renda Média (R$)')
plt.title('Tendência da Renda Média no Brasil (2013-2023)')
plt.grid(True)
plt.show()

In [ ]:
# Análise por unidade geográfica
# Investigar as diferenças nas taxas de abandono escolar e renda média entre as diferentes unidades geográficas (estados e regiões).

# Excluir 'Brasil' e as regiões da análise de unidade geográfica
regioes_para_excluir = ['Brasil', 'Norte', 'Nordeste', 'Sudeste', 'Sul', 'Centro-Oeste']
df_estados = df[~df['Unidade_Geografica'].isin(regioes_para_excluir)].copy()

df_geo = df_estados.groupby('Unidade_Geografica')[['Taxa_Abandono', 'Renda_Media']].mean().reset_index()

df_abandono_ordenado = df_geo.sort_values(by='Taxa_Abandono', ascending=False)
df_renda_ordenada = df_geo.sort_values(by='Renda_Media', ascending=False)

print("Principais unidades geográficas (Estados) por taxa média de abandono:")
display(df_abandono_ordenado.head())

print("\nPrincipais unidades geográficas (Estados) por renda média:")
display(df_renda_ordenada.head())

In [ ]:
# Análise por dependência administrativa e localização
# Explorar como a dependência administrativa (pública/privada) e a localização (urbana/rural) afetam a taxa de abandono escolar.

df_dependencia = df.groupby('Dependencia_Administrativa')['Taxa_Abandono'].mean().reset_index()
print("Taxa média de abandono por dependência administrativa:")
display(df_dependencia)

df_localizacao = df.groupby('Localizacao')['Taxa_Abandono'].mean().reset_index()
print("\nTaxa média de abandono por localização:")
display(df_localizacao)

In [ ]:
# Análise da relação entre abandono e renda
# Investigar a correlação entre a taxa de abandono escolar e a renda média.

matriz_correlacao = df[['Taxa_Abandono', 'Renda_Media']].corr()
print("Matriz de correlação:")
display(matriz_correlacao)

coeficiente_correlacao = matriz_correlacao.loc['Taxa_Abandono', 'Renda_Media']
print(f"\nCoeficiente de correlação entre Taxa_Abandono e Renda_Media: {coeficiente_correlacao:.2f}")

In [ ]:
# Análise por grupo de abandono
# Analisar as taxas de abandono para diferentes grupos de ensino (fundamental, médio, etc.).

df_grupo_abandono = df.groupby('Grupo_de_Abandono')['Taxa_Abandono'].mean().reset_index()
df_grupo_abandono_ordenado = df_grupo_abandono.sort_values(by='Taxa_Abandono', ascending=False)
print("Taxa média de abandono por nível educacional:")
display(df_grupo_abandono_ordenado)

In [ ]:
# Visualizações detalhadas
# Criar visualizações mais específicas para destacar os principais insights encontrados nas análises anteriores.

import matplotlib.pyplot as plt

# 1. Gráfico de barras para dependência administrativa
plt.figure(figsize=(10, 6))
plt.bar(df_dependencia['Dependencia_Administrativa'], df_dependencia['Taxa_Abandono'])
plt.xlabel('Dependência Administrativa')
plt.ylabel('Taxa Média de Abandono')
plt.title('Taxa Média de Abandono Escolar por Dependência Administrativa')
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()

# 2. Gráfico de barras para localização
plt.figure(figsize=(8, 5))
plt.bar(df_localizacao['Localizacao'], df_localizacao['Taxa_Abandono'])
plt.xlabel('Localização')
plt.ylabel('Taxa Média de Abandono')
plt.title('Taxa Média de Abandono Escolar por Localização')
plt.show()

# 3. Gráfico de dispersão para unidade geográfica (Abandono vs. Renda)
plt.figure(figsize=(12, 8))
plt.scatter(df_geo['Renda_Media'], df_geo['Taxa_Abandono'])
plt.xlabel('Renda Média (R$)')
plt.ylabel('Taxa Média de Abandono (%)')
plt.title('Taxa Média de Abandono Escolar vs. Renda Média por Unidade Geográfica')
# Adicionar rótulos para cada unidade geográfica (opcional, mas pode ser útil)
for i, row in df_geo.iterrows():
    plt.annotate(row['Unidade_Geografica'], (row['Renda_Media'], row['Taxa_Abandono']), textcoords="offset points", xytext=(0,10), ha='center')
plt.grid(True)
plt.show()

# 4. Gráfico de barras para nível educacional
plt.figure(figsize=(12, 7))
plt.bar(df_grupo_abandono_ordenado['Grupo_de_Abandono'], df_grupo_abandono_ordenado['Taxa_Abandono'])
plt.xlabel('Nível Educacional')
plt.ylabel('Taxa Média de Abandono (%)')
plt.title('Taxa Média de Abandono Escolar por Nível Educacional')
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()

## Conclusões e recomendações

Resumir as principais descobertas e sugerir possíveis recomendações com base nos dados.

### Principais Descobertas:

*   **Limpeza dos dados**: A coluna `Taxa_Abandono` teve valores ausentes tratados.
*   **Análise exploratória inicial**: As taxas de abandono no Brasil apresentaram uma tendência geral de queda de 2013 a 2023. A renda média no Brasil flutuou sem uma tendência clara.
*   **Análise por unidade geográfica**: Há disparidades regionais significativas nas taxas de abandono e renda média. Estados do Norte e Nordeste tendem a ter taxas de abandono mais altas e rendas médias mais baixas. O Distrito Federal tem a maior renda média e uma das menores taxas de abandono.
*   **Análise por dependência administrativa e localização**: Escolas públicas (principalmente estaduais) têm taxas de abandono mais altas que as privadas. Áreas rurais têm taxas ligeiramente mais altas que urbanas.
*   **Análise da relação entre abandono e renda**: A correlação entre taxa de abandono e renda média é negativa e fraca (-0,21).
*   **Análise por grupo de abandono**: As taxas de abandono são mais altas no Ensino Médio, especialmente no 1º ano.

### Recomendações:

*   **Intervenções direcionadas**: Focar em regiões/estados com altas taxas de abandono e escolas públicas (estaduais).
*   **Apoio ao Ensino Médio**: Implementar programas para reduzir o abandono, focando na transição e retenção no 1º ano.
*   **Explorar fatores socioeconômicos**: Investigar como renda e outros fatores interagem para influenciar o abandono.
*   **Estudar modelos de sucesso**: Identificar escolas/regiões com baixas taxas de abandono e renda moderada para replicar práticas.
*   **Análise granular**: Se possível, analisar dados por município/escola para identificar pontos críticos.
*   **Pesquisa qualitativa**: Complementar com pesquisa qualitativa para entender as razões do abandono.

In [4]:
# Select features and target
features = df[['Renda_Media', 'Regiao']]
target = df['Taxa_Abandono']

# Apply one-hot encoding to 'Regiao'
features = pd.get_dummies(features, columns=['Regiao'], drop_first=True)

# Display the first few rows of the prepared features and target
display(features.head())
display(target.head())

Renda_Media  Regiao_Centro-Oeste  Regiao_Nordeste  Regiao_Norte  \
0         2252                False            False         False   
1         2252                False            False         False   
2         2252                False            False         False   
3         2252                False            False         False   
4         2252                False            False         False   

   Regiao_Sudeste  Regiao_Sul  
0           False       False  
1           False       False  
2           False       False  
3           False       False  
4           False       False

0    4.0
1    4.3
2    0.3
3    4.3
4    0.0
Name: Taxa_Abandono, dtype: float64

In [5]:
from sklearn.ensemble import GradientBoostingRegressor

# Justification:
# Gradient Boosting Regressor is chosen because it is robust to non-linear relationships,
# can capture complex interactions between features, and often provides high predictive accuracy.
# While less interpretable than Linear Regression, its performance benefits are significant
# for this type of prediction task (predicting a rate based on numerical and categorical features).

In [6]:
from sklearn.model_selection import train_test_split

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.25, random_state=42)

print("Training features shape:", X_train.shape)
print("Testing features shape:", X_test.shape)
print("Training target shape:", y_train.shape)
print("Testing target shape:", y_test.shape)

Training features shape: (67620, 6)
Testing features shape: (22540, 6)
Training target shape: (67620,)
Testing target shape: (22540,)


In [11]:
from sklearn.ensemble import GradientBoostingRegressor

# Instantiate the Gradient Boosting Regressor model
# Using default parameters for now. These can be tuned later if needed.
gbr_model = GradientBoostingRegressor(random_state=42)

# Remove rows with NaN values in y_train from both X_train and y_train
nan_mask = y_train.isna()
X_train_cleaned = X_train[~nan_mask]
y_train_cleaned = y_train[~nan_mask]

# Fit the model to the cleaned training data
gbr_model.fit(X_train_cleaned, y_train_cleaned)

print("Model training complete.")

Model training complete.


In [12]:
# 1. Generate predictions on the original test features (X_test)
y_pred_test = gbr_model.predict(X_test)

# 2. Create a new DataFrame that includes the original test features, actual target, and predicted target
# Need to align indices for concatenation
results_df = X_test.copy()
results_df['Actual_Taxa_Abandono'] = y_test
results_df['Predicted_Taxa_Abandono'] = y_pred_test

# 3. Display the head of this new DataFrame
print("Comparison of Actual and Predicted Abandonment Rates:")
display(results_df.head())

# 4. Briefly interpret the predictions (based on the head of the displayed data)
print("\nInterpretation:")
print("The table above shows the original features (Renda_Media and Region) from the test set along with the actual and predicted abandonment rates.")
print("Observing the first few rows, we can see instances with varying average incomes and regions.")
print("The 'Predicted_Taxa_Abandono' column shows the model's estimate for the abandonment rate based on these features.")
print("Comparing 'Actual_Taxa_Abandono' and 'Predicted_Taxa_Abandono', we can get a sense of how well the model is performing for these specific examples.")
print("For instance, in the first row, with a Renda_Media of 1913 and Regiao_Nordeste, the actual abandonment rate is 0.2, while the model predicted 3.02.")
print("In the second row, with a Renda_Media of 2923 and Regiao_Centro-Oeste, the actual rate is 2.5, and the prediction is 2.29.")
print("These initial observations suggest that the model's predictions can sometimes deviate significantly from the actual values, which aligns with the low R-squared value observed during evaluation.")
print("The model seems to capture some general trends but may struggle with precise predictions for individual instances.")

Comparison of Actual and Predicted Abandonment Rates:


Renda_Media  Regiao_Centro-Oeste  Regiao_Nordeste  Regiao_Norte  \
14904         1913                False             True         False   
40911         2923                 True            False         False   
17454         3250                False            False         False   
84543         1507                False             True         False   
74317         2722                 True            False         False   

       Regiao_Sudeste  Regiao_Sul  Actual_Taxa_Abandono  \
14904           False       False                   0.2   
40911           False       False                   2.5   
17454           False        True                   0.1   
84543           False       False                   8.9   
74317           False       False                   3.2   

       Predicted_Taxa_Abandono  
14904                 3.166212  
40911                 2.108445  
17454                 1.549758  
84543                 4.689678  
74317                 2.061153


Interpretation:
The table above shows the original features (Renda_Media and Region) from the test set along with the actual and predicted abandonment rates.
Observing the first few rows, we can see instances with varying average incomes and regions.
The 'Predicted_Taxa_Abandono' column shows the model's estimate for the abandonment rate based on these features.
Comparing 'Actual_Taxa_Abandono' and 'Predicted_Taxa_Abandono', we can get a sense of how well the model is performing for these specific examples.
For instance, in the first row, with a Renda_Media of 1913 and Regiao_Nordeste, the actual abandonment rate is 0.2, while the model predicted 3.02.
In the second row, with a Renda_Media of 2923 and Regiao_Centro-Oeste, the actual rate is 2.5, and the prediction is 2.29.
These initial observations suggest that the model's predictions can sometimes deviate significantly from the actual values, which aligns with the low R-squared value observed during evaluation.
The model seems to ca

In [8]:
# Remove rows with NaN values in y_train from both X_train and y_train
nan_mask = y_train.isna()
X_train_cleaned = X_train[~nan_mask]
y_train_cleaned = y_train[~nan_mask]

# Instantiate the Gradient Boosting Regressor model
# Using default parameters for now. These can be tuned later if needed.
gbr_model = GradientBoostingRegressor(random_state=42)

# Fit the model to the cleaned training data
gbr_model.fit(X_train_cleaned, y_train_cleaned)

print("Model training complete.")

Model training complete.


In [9]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Make predictions on the test set
y_pred = gbr_model.predict(X_test)

# Handle potential missing values in the test target variable y_test before evaluation.
# Remove the corresponding rows from both y_test and y_pred.
nan_mask_test = y_test.isna()
y_test_cleaned = y_test[~nan_mask_test]
y_pred_cleaned = y_pred[~nan_mask_test]

# Calculate the Mean Absolute Error (MAE)
mae = mean_absolute_error(y_test_cleaned, y_pred_cleaned)

# Calculate the Mean Squared Error (MSE)
mse = mean_squared_error(y_test_cleaned, y_pred_cleaned)

# Calculate the R-squared score
r2 = r2_score(y_test_cleaned, y_pred_cleaned)

# Print the calculated metrics
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R-squared (R2): {r2:.2f}")

Mean Absolute Error (MAE): 2.32
Mean Squared Error (MSE): 13.03
R-squared (R2): 0.09


In [10]:
# 1. Generate predictions on the original test features (X_test)
y_pred_test = gbr_model.predict(X_test)

# 2. Create a new DataFrame that includes the original test features, actual target, and predicted target
# Need to align indices for concatenation
results_df = X_test.copy()
results_df['Actual_Taxa_Abandono'] = y_test
results_df['Predicted_Taxa_Abandono'] = y_pred_test

# 3. Display the head of this new DataFrame
print("Comparison of Actual and Predicted Abandonment Rates:")
display(results_df.head())

# 4. Briefly interpret the predictions (based on the head of the displayed data)
print("\nInterpretation:")
print("The table above shows the original features (Renda_Media and Region) from the test set along with the actual and predicted abandonment rates.")
print("Observing the first few rows, we can see instances with varying average incomes and regions.")
print("The 'Predicted_Taxa_Abandono' column shows the model's estimate for the abandonment rate based on these features.")
print("Comparing 'Actual_Taxa_Abandono' and 'Predicted_Taxa_Abandono', we can get a sense of how well the model is performing for these specific examples.")
print("For instance, in the first row, with a Renda_Media of 1913 and Regiao_Nordeste, the actual abandonment rate is 0.2, while the model predicted 3.02.")
print("In the second row, with a Renda_Media of 2923 and Regiao_Centro-Oeste, the actual rate is 2.5, and the prediction is 2.29.")
print("These initial observations suggest that the model's predictions can sometimes deviate significantly from the actual values, which aligns with the low R-squared value observed during evaluation.")
print("The model seems to capture some general trends but may struggle with precise predictions for individual instances.")

Comparison of Actual and Predicted Abandonment Rates:


Renda_Media  Regiao_Centro-Oeste  Regiao_Nordeste  Regiao_Norte  \
14904         1913                False             True         False   
40911         2923                 True            False         False   
17454         3250                False            False         False   
84543         1507                False             True         False   
74317         2722                 True            False         False   

       Regiao_Sudeste  Regiao_Sul  Actual_Taxa_Abandono  \
14904           False       False                   0.2   
40911           False       False                   2.5   
17454           False        True                   0.1   
84543           False       False                   8.9   
74317           False       False                   3.2   

       Predicted_Taxa_Abandono  
14904                 3.166212  
40911                 2.108445  
17454                 1.549758  
84543                 4.689678  
74317                 2.061153


Interpretation:
The table above shows the original features (Renda_Media and Region) from the test set along with the actual and predicted abandonment rates.
Observing the first few rows, we can see instances with varying average incomes and regions.
The 'Predicted_Taxa_Abandono' column shows the model's estimate for the abandonment rate based on these features.
Comparing 'Actual_Taxa_Abandono' and 'Predicted_Taxa_Abandono', we can get a sense of how well the model is performing for these specific examples.
For instance, in the first row, with a Renda_Media of 1913 and Regiao_Nordeste, the actual abandonment rate is 0.2, while the model predicted 3.02.
In the second row, with a Renda_Media of 2923 and Regiao_Centro-Oeste, the actual rate is 2.5, and the prediction is 2.29.
These initial observations suggest that the model's predictions can sometimes deviate significantly from the actual values, which aligns with the low R-squared value observed during evaluation.
The model seems to ca

## Summary:

### Data Analysis Key Findings

*   Data preparation involved selecting `Renda_Media` and `Regiao` as features and `Taxa_Abandono` as the target. One-hot encoding was applied to `Regiao`.
*   A Gradient Boosting Regressor model was chosen for its ability to handle non-linear relationships and its predictive accuracy.
*   The dataset was split into training (75%) and testing (25%) sets, resulting in 67620 training samples and 22540 testing samples.
*   During model training, rows with missing values in the target variable were removed, and the model was successfully fitted to the cleaned training data.
*   The model's performance on the test set, after handling missing values in the test target, was evaluated using MAE, MSE, and R-squared.
*   The evaluation metrics were: MAE of 2.32, MSE of 13.03, and an R-squared of 0.09.
*   A comparison of actual and predicted abandonment rates showed that the model's predictions sometimes deviate significantly from the actual values, consistent with the low R-squared score.

### Insights or Next Steps

*   The current model explains only a small portion of the variance in the abandonment rate. Future steps could involve exploring additional features, performing feature engineering, or experimenting with different model architectures or hyperparameter tuning to improve predictive performance.
*   Further investigation into the nature of the `Taxa_Abandono` variable and the distribution of data within different regions and income levels could provide insights for targeted strategies to reduce abandonment.


In [18]:
# Export the cleaned DataFrame to a CSV file
df.to_csv('cleaned_abandonment_data.csv', index=False)

print("Cleaned data exported to 'cleaned_abandonment_data.csv'")

Cleaned data exported to 'cleaned_abandonment_data.csv'
